[reland][bc-breaking] reference option for linear produce a pattern instead of reference linear module (#61892)

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62277

Summary:

This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Imported from OSS

Reviewed By: vkuzo

Differential Revision: [D29941079](https://our.internmc.facebook.com/intern/diff/D29941079)

jerryzh168 committed:
[reland][bc-breaking] reference option for linear produce a pattern instead of reference linear module (#61892)

Summary:

This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Imported from OSS

Reviewed By: vkuzo

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29941079).

ejguan commented:
It seems mypy still complains about
torch/nn/intrinsic/qat/modules/linear_relu.py:44:23: error: Incompatible types in assignment (expression has type "None", variable has type "Parameter")  [assignment]

jerryzh168 committed:
Update on "[reland][bc-breaking] reference option for linear produce a pattern instead of reference linear module (#61892)"


Summary:

This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Imported from OSS

Reviewed By: vkuzo

Differential Revision: [D29941079](https://our.internmc.facebook.com/intern/diff/D29941079)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29941079).

ejguan reviewed (approved):
Accept the updated PR to address the mypy Error.
